receiver/aerospikereceiver/scraper.go

```
@@ -0,0 +1,217 @@
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+)
+
+// aerospikeReceiver is a metrics receiver using the Aerospike interface to collect
+type aerospikeReceiver struct {
+	params   component.ReceiverCreateSettings
+	config   *Config
+	consumer consumer.Metrics
+	host     string // host/IP of configured Aerospike node
+	port     int    // port of configured Aerospike node
+	mb       *metadata.MetricsBuilder
+	logger   *zap.Logger
+}
+
+// newAerospikeReceiver creates a new aerospikeReceiver connected to the endpoint provided in cfg
+//
+// If the host or port can't be parsed from endpoint, an error is returned.
+func newAerospikeReceiver(params component.ReceiverCreateSettings, cfg *Config, consumer consumer.Metrics) (*aerospikeReceiver, error) {
+	host, portStr, err := net.SplitHostPort(cfg.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("endpoint: %w", err)
+	}
+
+	port, err := strconv.ParseInt(portStr, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("port: %w", err)
+	}
+
+	return &aerospikeReceiver{
+		params:   params,
+		config:   cfg,
+		consumer: consumer,
+		host:     host,
+		port:     int(port),
+		mb:       metadata.NewMetricsBuilder(cfg.Metrics),
+		logger:   params.Logger.Named("aerospikereceiver"),
+	}, nil
+}
+
+// scrape scrapes both Node and Namespace metrics from the provided Aerospike node.
+// If CollectClusterMetrics is true, it then scrapes every discovered node
+func (r *aerospikeReceiver) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	errs := &scrapererror.ScrapeErrors{}
+	now := pcommon.NewTimestampFromTime(time.Now().UTC())
+	client, err := newASClient(r.host, r.port, r.config.Timeout)
+	if err != nil {
+		r.logger.Warn(fmt.Sprintf("failed to connect: %s", err.Error()))
+		return r.mb.Emit(), nil
+	}
```

We should return the error here.